components/range_cache_memory_engine/src/range_manager.rs

```diff
@@ -199,127 +318,237 @@ impl RangeManager {
+            .range::<[u8], (Bound<&[u8]>, Bound<&[u8]>)>((Excluded(&range.start), Unbounded))
+            .next();
+        if let Some((_key, id)) = entry {
+            let meta = &self.regions[id];
+            if meta.range.start < range.end {
+                return true;
+            }
+        }
+        false
     }
 
-    fn overlap_with_pending_range(&self, range: &CacheRange) -> bool {
-        self.pending_ranges.iter().any(|r| r.overlaps(range))
-            || self
-                .pending_ranges_loading_data
-                .iter()
-                .any(|(r, ..)| r.overlaps(range))
+    /// `check_overlap_with_region` check of there are regions overlap with
+    /// target region. If there are regions with `pending` state and whose
+    /// epoch verion is smaller than target region, the pending regions will
+    /// be removed first.
+    fn check_overlap_with_region(&mut self, region: &Region) -> Option<RegionState> {
+        let region_range = CacheRange::from_region(region);
+        let mut removed_region = vec![];
+        let mut overlapped_region_state = None;
+        self.iter_overlapped_regions(&region_range, |region_meta| {
+            if region_meta.state != RegionState::Pending
+                || region_meta.region.get_region_epoch().version
+                    >= region.get_region_epoch().version
+            {
+                tikv_util::warn!("load region overlaps with existing region"; "region" => ?region, "exist_meta" => ?region_meta);
+                overlapped_region_state = Some(region_meta.state);
+                return false;
+            }
+            // pending region with out-dated epoch, should be removed.
+            removed_region.push(region_meta.region.id);
+            true
+        });
+        for id in removed_region {
+            self.remove_region(id);
+        }
+        overlapped_region_state
     }
 
     // Acquire a snapshot of the `range` with `read_ts`. If the range is not
     // accessable, None will be returned. Otherwise, the range id will be returned.
-    pub(crate) fn range_snapshot(
+    pub(crate) fn region_snapshot(
         &mut self,
-        range: &CacheRange,
+        region_id: u64,
+        region_epoch: u64,
         read_ts: u64,
-    ) -> result::Result<u64, FailedReason> {
-        let Some(range_key) = self
-            .ranges
-            .keys()
-            .find(|&r| r.contains_range(range))
-            .cloned()
-        else {
+    ) -> result::Result<(), FailedReason> {
+        let Some(meta) = self.regions.get_mut(&region_id) else {
             return Err(FailedReason::NotCached);
         };
-        let meta = self.ranges.get_mut(&range_key).unwrap();
+
+        if meta.state != RegionState::Active {
+            return Err(FailedReason::NotCached);
+        }
+
+        if meta.region.get_region_epoch().version != region_epoch {
+            return Err(FailedReason::EpochNotMatch);
+        }
 
         if read_ts <= meta.safe_point {
```

Comment:
I think it still can serve the read request if read_ts == safe_point.

Reply:
https://github.com/tikv/tikv/blob/26c20c7923fb868cd3976fb8d36d738d4fff4abb/components/range_cache_memory_engine/src/background.rs#L1374-L1409

From the current implementation, KVs' commit_ts == safe_point will also be removed